README.md
# mybigdata

[![Build Status](https://travis-ci.org/supermy/kafka-spark-redis.svg?branch=master)](https://github.com/supermy/kafka-spark-redis)

## 简介 
*  大数据实时流的计算框架，kafka+spark+redis
*  确保 spark所使用的scala版本与你系统scala的版本一致



## 特点
* 实时计算
* 窗口框架计算
* Spark 调用Redis采用单例+pipeline ，性能比普通模式提升10倍


## 适用场景

时间片段数量统计
时间片段筛选


## 优势：

* kafka 大数据量传输优势
* spark 时间窗的灵活计算
* redis 基于内存的定制计算

## 劣势：

 Redis 最大存储量的限制


## 快速试用
    1.  需要 jdk8环境；
    2.  配置 kafka 环境；
    3.  运行示例1
    4.  运行示例2



## kafka 环境配置 

### 安装
```aidl
    brew install kafka redis 
```

### 配置及性能测试
```
1.先启动zookeeper服务，端口确保2181.

2.kafka-server-start.sh /usr/local/etc//kafka/server.properties

3.创建一个主题
    创建，
        kafka-topics.sh --create --zookeeper localhost:2181 --replication-factor 1 --partitions 1 --topic test 
    获取，
        kafka-topics.sh --list --zookeeper localhost:2181

4.发送一些消息: 在控制台输入消息
    kafka-console-producer.sh --broker-list localhost:9092 --topic test 

5.启动一个消费者：观察在数据台显示消息
    kafka-console-consumer.sh --zookeeper localhost:2181 --topic test --from-beginning


Topic的分区和复制
1. 创建debugo01，这个topic分区数为3，复制为1（不复制）。该topic跨越全部broker。下面管理命令在任意kafka节点上执行即可
    kafka-topics.sh --create --zookeeper debugo01,debugo02,debugo03 --replication-factor 1 --partitions 3 --topic debugo01
    kafka-topics.sh --create --zookeeper localhost --replication-factor 1 --partitions 3 --topic debugo01

2. 创建debugo02，这个topic分区数为1，复制为3（每个主机都有一份）。该topic跨越全部broker。下面管理命令在任意kafka节点上执行即可
    需要三台主机
    kafka-topics.sh --create --zookeeper debugo01,debugo02,debugo03 --replication-factor 3 --partitions 1 --topic debugo02
    
3. 列出topic信息
    kafka-topics.sh --list --zookeeper localhost:2181
    
4. 列出topic描述信息
    kafka-topics.sh --describe --zookeeper localhost:2181 --topic debugo01
    
5. 检查log目录，对于topic debugo01，debugo01为0号分区，debugo02为1号分区。而topic debugo02则复制了3份，都为0号分区
    
6. 下面topic debugo03，replication-factor为2，partition为3.那么broker id为1的debugo01会如下面describe所示，保存0号分区和1号分区。
    而0号分区的repica leader为broker id = 3，包含3和1两个replicas。
        kafka-topics.sh --create --zookeeper debugo01,debugo02,debugo03 --replication-factor 2 --partitions 3 --topic debugo03
    
    kafka-topics.sh --describe --zookeeper localhost:2181 --topic debugo03
    ll /var/kafka/debugo03*
    
    消息的产生和消费
    kafka-console-producer.sh --broker-list debugo01:9092 --topic debugo03
    kafka-console-consumer.sh --zookeeper debugo01:2181 --from-beginning --topic debugo03
    

7.性能测试
    4线程
    生产数据：50w/12sec 4w/sec
    消费数据：150/27sec 5w/sec
    下面使用perf命令来测试几个topic的性能，需要先下载kafka-perf_2.10-0.8.1.1.jar，并拷贝到kafka/libs下面。
    50W条消息，每条1000字节，batch大小1000，topic为debugo01，4个线程（message size设置太大需要调整相关参数，否则容易OOM）。只用了13秒完成，kafka在多分区支持下吞吐量是非常给力的。
    
        kafka-producer-perf-test.sh --messages 500000 --message-size 1000  --batch-size 1000 --topics debugo01 --threads 4 --broker-list debugo01:9092,debugo02:9092,debugo03:9092
    
    同样的参数测试debugo02, 由于但分区加复制（replicas-factor=3），用时39秒。所以，适当加大partition数量和broker相关线程数量会极大的提高性能。
        kafka-producer-perf-test.sh --messages 500000 --message-size 1000  --batch-size 1000 --topics debugo02 --threads 4 --broker-list debugo01:9092,debugo02:9092,debugo03:9092
    
    同样的参数测试debugo03，用时30秒。
        kafka-producer-perf-test.sh --messages 500000 --message-size 1000  --batch-size 1000 --topics debugo03 --threads 4 --broker-list debugo01:9092,debugo02:9092,debugo03:9092    
    
        kafka-consumer-perf-test.sh --zookeeper debugo01,debugo02,debugo03 --messages 500000 --topic debugo01 --threads 3
        kafka-consumer-perf-test.sh --zookeeper debugo01,debugo02,debugo03 --messages 500000 --topic debugo02 --threads 3
        kafka-consumer-perf-test.sh --zookeeper debugo01,debugo02,debugo03 --messages 500000 --topic debugo03 --threads 3
    
    单机测试
        kafka-producer-perf-test.sh --messages 500000 --message-size 1000  --batch-size 1000 --topics debugo01 --threads 4 --broker-list localhost:9092
        kafka-consumer-perf-test.sh --zookeeper localhost --messages 500000 --topic debugo01 --threads 3


```

## 编译运行

```aidl
    mvn scala:compile
    
```
## 使用

### 示例1：生产数据，KafkaWordCountProducer 消费数据，KafkaWordCount

```aidl
        spark-streaming-kafka 包：支持向 kafka 传送数据
            //topic 与 随机数 构造消息
            val message = new ProducerRecord[String, String](topic, null, str)
            //发送消息到 kafka
            producer.send(message)
        
        

        运行时编辑configuration,
        (1)KafkaWordCountProducer
        选择KafkaWordCount.scala中的KafkaWordCountProducer方法
        VM options 设置为:-Dspark.master=local
        设置程序输入参数,Program arguments: localhost:9092 test 3 5
        
```

```aidl
        从 zookeeper 获取地址消费数据，此方式已被废弃
        val lines = KafkaUtils.createStream(ssc, zkQuorum, group, topicpMap).map(_._2) //创建流并且获取数据

```

### 示例2：生产数据，KafkaEventProducer$ 消费数据，UserClickCountAnalytics$
```aidl
    生产数据
          val producer = new Producer[String, String](kafkaConfig)
          producer.send(new KeyedMessage[String, String](topic, event.toString))

    消费数据，目前采用的方式快速且好用资源较少
          val kafkaStream = KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](ssc, kafkaParams, topics)

```

### 示例3：打包提交版本到生产环境

```aidl
    1. source /etc/profile  
    2.   
    3. grapher=`ps -ef | grep spark |grep SparkStreaming.jar | awk '{print $2}'`  
    4. echo $grapher  
    5.   
    6. kill -9 $grapher  
    7.   
    8. nohup /opt/modules/spark/bin/spark-submit \  
    9. --master spark://127.0.0.1:7077 \  
    10. --driver-memory 3g \  
    11. --executor-memory 3g \  
    12. --total-executor-cores 24 \  
    13. --conf spark.ui.port=56689  \  
    14. --jars /opt/bin/sparkJars/kafka_2.10-0.8.2.1.jar,/opt/bin/sparkJars/spark-streaming-kafka_2.10-1.4.1.jar,/opt/bin/sparkJars/metrics-core-2.2.0.jar,/opt/bin/sparkJars/mysql-connector-java-5.1.26-bin.jar,/opt/bin/sparkJars/spark-streaming-  
    15. kafka_2.10-1.4.1.jar \  
    16. --class com.hexun.streaming.StockCntSumKafkaLPcnt \  
    17. /opt/bin/UDF/SparkStreaming.jar \  
    18.  >/opt/bin/initservice/stock.log 2>&1 & \ 

```

CHANGELOG.md
### 20171030
    pipeline 模式redis性能提升10倍

### 2017-10-28
    整理文档
    
### 2017-10-26
    
    从2015年1月的笔记中整理脚本为项目代码
    

src/main/java/KafkaWordCount.scala
import org.apache.spark.SparkConf
import org.apache.spark.streaming._
import org.apache.spark.streaming.kafka._

/**
  * 消费数据
  *
  * <!-- 注意:这里scala的版本要和自己电脑上安装的scala一致，而且要注意spark对应的版本是否支持scala对应的版本。否则会存在版本冲突问题 -->
  * 步骤2：运行KafkaWordCount
  * args：localhost:2181 test-consumer-group test 1
  * localhost:2181表示zookeeper的监听地址，test-consumer-group表示consumer-group的名称，test表示topic，1表示线程数。
  *
  *
  * 基于direct的方式，使用kafka的简单api，Spark Streaming自己就负责追踪消费的offset，并保存在checkpoint中。Spark自己一定是同步的，因此可以保证数据是消费一次且仅消费一次。
  * 1. val topics = Set("teststreaming")
2. val brokers = "bdc46.hexun.com:9092,bdc53.hexun.com:9092,bdc54.hexun.com:9092"
3. val kafkaParams = Map[String, String]("metadata.broker.list" -> brokers, "serializer.class" -> "kafka.serializer.StringEncoder")
4. // Create a direct stream
5. val kafkaStream = KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](ssc, kafkaParams, topics)
6. val events = kafkaStream.flatMap(line => {
7. Some(line.toString())
8. })
  新api的Direct应该是以后的首选方式
  *
  * Created by moyong on 2017/10/27.
  */
object KafkaWordCount {
  def main(args: Array[String]) {
    if (args.length < 4) {
      System.err.println("Usage: KafkaWordCount    ")
      System.exit(1)
    }


    val Array(zkQuorum, group, topics, numThreads) = args
    val sparkConf = new SparkConf().setAppName("KafkaWordCount")  //spark 配置
    val ssc =  new StreamingContext(sparkConf, Seconds(2)) //spark 流配置
    ssc.checkpoint("checkpoint") //配置检查点

    val topicpMap = topics.split(",").map((_,numThreads.toInt)).toMap //获取 topic 列表
    val lines = KafkaUtils.createStream(ssc, zkQuorum, group, topicpMap).map(_._2) //创建流并且获取数据
    val words = lines.flatMap(_.split(" "))  //拆分数据
    val wordCounts = words.map(x => (x, 1L))  //构造 map 值为 key,数据为1
      .reduceByKeyAndWindow(_ + _, _ - _, Minutes(10), Seconds(2), 2) //进行窗口统计，按值排序
//      .reduceByKeyAndWindow(_ + _, _ - _, Seconds(10), Seconds(2), 2) //进行窗口统计，按值排序

    wordCounts.print() //打印统计数据

    ssc.start()
    ssc.awaitTermination()
  }
}

src/main/java/KafkaWordCountProducer.scala
import java.util  
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}

/**
  * kafka 数据生产
  *  (1)KafkaWordCountProducer
    选择KafkaWordCount.scala中的KafkaWordCountProducer方法
    VM options 设置为:-Dspark.master=local
    设置程序输入参数,Program arguments: localhost:9092 test 3 5
  *
  */
object KafkaWordCountProducer {  
  def main(args: Array[String]) {  
//    val topic = "test"
//    val brokers = "localhost:9092"
//    val messagesPerSec= 3 //每秒发送几条信息
//    val wordsPerMessage = 5 //一条信息包括多少个单词

    if (args.length < 4) {
      System.err.println("Usage: KafkaWordCountProducer <metadataBrokerList> <topic> " +
        "<messagesPerSec> <wordsPerMessage>")
      System.exit(1)
    }

    val Array(brokers, topic, messagesPerSec, wordsPerMessage) = args

    // Zookeeper connection properties  配置 zookeeper 参数
    val props = new util.HashMap[String, Object]()  
    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokers)  
    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,  
      "org.apache.kafka.common.serialization.StringSerializer")  
    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,  
      "org.apache.kafka.common.serialization.StringSerializer")  
      
    val producer = new KafkaProducer[String, String](props)  
    // Send some messages  
    while(true) {  
      (0 to messagesPerSec.toInt).foreach { messageNum =>
        //参数5个随机数，用空格分开
        val str = (0 to wordsPerMessage.toInt).map(x => scala.util.Random.nextInt().toString).mkString(" ")
        //topic 与 随机数 构造消息
        val message = new ProducerRecord[String, String](topic, null, str)
        //发送消息到 kafka
        producer.send(message)
        //打印消息
        println(message)
      }
      Thread.sleep(10)

    }  
  }  
}  

src/main/java/UserClickCountAnalytics.scala
import com.google.gson.{Gson, JsonObject, JsonParser}
import org.apache.spark.SparkConf
import org.apache.spark.streaming.Seconds
import org.apache.spark.streaming.StreamingContext
import org.apache.spark.streaming.kafka.KafkaUtils
import kafka.serializer.StringDecoder
import org.apache.commons.pool2.impl.GenericObjectPoolConfig
import redis.clients.jedis.{Jedis, JedisPool, JedisPoolConfig}



/**
  * Created by moyong on 2017/10/27.
  */
object UserClickCountAnalytics {

  def main(args: Array[String]): Unit = {
    var masterUrl = "local[1]"
    if (args.length > 0) {
      masterUrl = args(0)
    }

    // Create a StreamingContext with the given master URL
    val conf = new SparkConf().setMaster(masterUrl).setAppName("UserClickCountStat")
    val ssc = new StreamingContext(conf, Seconds(5))  //5秒间隔

    // Kafka configurations
    val topics = Set("user_events")
    val brokers = "127.0.0.1:9092,127.0.0.1:9092"
    val kafkaParams = Map[String, String](
      "metadata.broker.list" -> brokers, "serializer.class" -> "kafka.serializer.StringEncoder")

    val dbIndex = 1
    val clickHashKey = "app::users::click"

    // Create a direct stream
    val kafkaStream = KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](ssc, kafkaParams, topics)

    val events = kafkaStream.flatMap(line => {
      val data =  new JsonParser().parse(line._2).getAsJsonObject()
      Some(data)
    })

    // Compute user click times
    val userClicks = events.map(x => (x.get("uid").getAsString, x.get("click_count").getAsInt)).reduceByKey(_ + _)
    userClicks.foreachRDD(rdd => {
      rdd.foreachPartition(partitionOfRecords => {

        //集群模式+pipeline

        /**
          * Internal Redis client for managing Redis connection {@link Jedis} based on {@link RedisPool}
          */
        object InternalRedisClient extends Serializable {

          @transient private var pool: JedisPool = null

          def makePool(redisHost: String, redisPort: Int, redisTimeout: Int,
                       maxTotal: Int, maxIdle: Int, minIdle: Int): Unit = {
            makePool(redisHost, redisPort, redisTimeout, maxTotal, maxIdle, minIdle, true, false, 10000)
          }

          //scala 的单例模式
          def makePool(redisHost: String, redisPort: Int, redisTimeout: Int,
                       maxTotal: Int, maxIdle: Int, minIdle: Int, testOnBorrow: Boolean,

                       testOnReturn: Boolean, maxWaitMillis: Long): Unit = {

            if(pool == null) {

              val poolConfig = new GenericObjectPoolConfig()

              poolConfig.setMaxTotal(maxTotal)

              poolConfig.setMaxIdle(maxIdle)

              poolConfig.setMinIdle(minIdle)

              poolConfig.setTestOnBorrow(testOnBorrow)
              poolConfig.setTestOnReturn(testOnReturn)
              poolConfig.setMaxWaitMillis(maxWaitMillis)
              pool = new JedisPool(poolConfig, redisHost, redisPort, redisTimeout)
              val hook = new Thread{
                override def run = pool.destroy()
              }
              sys.addShutdownHook(hook.run)
            }
          }
          def getPool: JedisPool = {
            assert(pool != null)
            pool
          }
        }
        // Redis configurations
        val maxTotal = 10
        val maxIdle = 10
        val minIdle = 1
        val redisHost = "127.0.0.1"
        val redisPort = 6379
        val redisTimeout = 30000
        val dbIndex = 1

        InternalRedisClient.makePool(redisHost, redisPort, redisTimeout, maxTotal, maxIdle, minIdle)

        val jedis =InternalRedisClient.getPool.getResource
        jedis.select(dbIndex)

        val p = jedis.pipelined()


        partitionOfRecords.foreach(pair => {
//          val uid = pair._1
//          val clickCount = pair._2
//
//          //本地模式
//          val jedis = RedisClient.pool.getResource
//          jedis.select(dbIndex)
//          jedis.hincrBy(clickHashKey, uid, clickCount)
//          RedisClient.pool.returnResource(jedis)

          val uid = pair._1
          val clickCount = pair._2
          jedis.hincrBy(clickHashKey, uid, clickCount)


        })


        p.sync()
        InternalRedisClient.getPool.returnResource(jedis)

      })
    })

    ssc.start()
    ssc.awaitTermination()

  }
}


src/main/java/KafkaEventProducer.scala
/**
  * Created by moyong on 2017/10/27.
  */

import java.util.Properties
import scala.util.Properties
import org.codehaus.jettison.json.JSONObject
import kafka.javaapi.producer.Producer
import kafka.producer.KeyedMessage
import kafka.producer.KeyedMessage
import kafka.producer.ProducerConfig
import scala.util.Random

/**
  * uid：用户编号
event_time：事件发生时间戳
os_type：手机App操作系统类型
click_count：点击次数
  */
object KafkaEventProducer {

  private val users = Array(
    "4A4D769EB9679C054DE81B973ED5D768", "8dfeb5aaafc027d89349ac9a20b3930f",
    "011BBF43B89BFBF266C865DF0397AA71", "f2a8474bf7bd94f0aabbd4cdd2c06dcf",
    "068b746ed4620d25e26055a9f804385f", "97edfc08311c70143401745a03a50706",
    "d7f141563005d1b5d0d3dd30138f3f62", "c8ee90aade1671a21336c721512b817a",
    "6b67c8c700427dee7552f81f3228c927", "a95f22eabc4fd4b580c011a3161a9d9d")

  private val random = new Random()

  private var pointer = -1

  def getUserID() : String = {
    pointer = pointer + 1
    if(pointer >= users.length) {
      pointer = 0
      users(pointer)
    } else {
      users(pointer)
    }
  }

  def click() : Double = {
    random.nextInt(10)
  }

  // bin/kafka-topics.sh --zookeeper zk1:2181,zk2:2181,zk3:2181/kafka --create --topic user_events --replication-factor 2 --partitions 2
  // bin/kafka-topics.sh --zookeeper zk1:2181,zk2:2181,zk3:2181/kafka --list
  // bin/kafka-topics.sh --zookeeper zk1:2181,zk2:2181,zk3:2181/kafka --describe user_events
  // bin/kafka-console-consumer.sh --zookeeper zk1:2181,zk2:2181,zk3:22181/kafka --topic test_json_basis_event --from-beginning
  def main(args: Array[String]): Unit = {
    val topic = "user_events"
    val brokers = "127.0.0.1:9092,127.0.0.1:9092"
    val props = new Properties()
    props.put("metadata.broker.list", brokers)
    props.put("serializer.class", "kafka.serializer.StringEncoder")

    val kafkaConfig = new ProducerConfig(props)
    val producer = new Producer[String, String](kafkaConfig)

    while(true) {
      // prepare event data
      val event = new JSONObject()
      event
        .put("uid", getUserID)
        .put("event_time", System.currentTimeMillis.toString)
        .put("os_type", "Android")
        .put("click_count", click)

      // produce event message
      producer.send(new KeyedMessage[String, String](topic, event.toString))
      println("Message sent: " + event)

      Thread.sleep(200)
    }
  }
}

